pkg/apis/core/validation/validation.go

```diff
@@ -3120,7 +3120,7 @@ func ValidatePod(pod *core.Pod) field.ErrorList {
 			}
 		}
 	}
-	if len(hugePageResources) > 1 {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.HugePageStorageMediumSize) && len(hugePageResources) > 1 {
```

Comment:
This doesn't account for a version N API server (with the feature off) validating an update to a pod that was allowed in by a version N+1 API server (with the feature on).

We always need to tolerate a no-op update to an existing object, regardless of feature-enablement.

The following scenario is important:
1. cluster has three API servers at version N, HugePageStorageMediumSize feature is off.
2. cluster upgrades one API server to version N+1, HugePageStorageMediumSize feature is on on that version.
3. user creates a pod using more than one hugepage resource against the N+1 API server (which passes validation)
4. with this PR's current state, that pod could not be updated via the version N API server.

A similar scenario was discussed in https://github.com/kubernetes/kubernetes/pull/82820#issuecomment-571341191 (see the next several comments as well)

The resolution looked like:
* telling validation whether multiple hugepages resources were allowed (rather than checking feature gates in validation)
* on create, allow multiple hugepages if the feature is enabled
* on update, allow multiple hugepages if the feature is enabled OR if the existing object already has multiple hugepages specified

Reply:
Also see https://github.com/kubernetes/kubernetes/issues/72651 for why it is generally better not to check feature gates directly in validation.go (because validation often doesn't have the context to know about data in an already persisted object)

Reply:
@liggitt Thank you for the review and detailed instructions! Should be done now. Pleaese review again.